File: src/main/java/com/timepath/hl2/hudeditor/EditorMenuBar.java
package com.timepath.hl2.hudeditor;

import com.timepath.plaf.OS;
import com.timepath.vgui.Element;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.KeyEvent;

/**
 * @author TimePath
 */
class EditorMenuBar extends JMenuBar {

    HUDEditor hudEditor;
    JMenuItem newItem, openItem, saveItem, saveAsItem, reloadItem, closeItem, exitItem;
    JMenuItem undoItem, redoItem, cutItem, copyItem, pasteItem, deleteItem, selectAllItem, preferencesItem;
    JMenuItem resolutionItem, previewItem;
    int mod = Toolkit.getDefaultToolkit().getMenuShortcutKeyMask();
    JMenuItem aboutItem;

    EditorMenuBar(final HUDEditor hudEditor) {
        this.hudEditor = hudEditor;
        add(new JMenu(Main.getString("File")) {{
            setMnemonic(KeyEvent.VK_F);
            add(newItem = new JMenuItem(new CustomAction(Main.getString("New"), KeyEvent.VK_N, ks(KeyEvent.VK_N, mod)) {
                @Override
                void action(ActionEvent e) {
                    throw new UnsupportedOperationException("Not supported yet.");
                }
            }));
            newItem.setEnabled(false);
            add(openItem = new JMenuItem(new CustomAction("Open", KeyEvent.VK_O, ks(KeyEvent.VK_O, mod)) {
                @Override
                void action(ActionEvent e) {
                    hudEditor.locateHudDirectory();
                }
            }));
            addSeparator();
            closeItem = new JMenuItem(new CustomAction("Close", KeyEvent.VK_C, ks(KeyEvent.VK_W, mod)) {
                @Override
                void action(ActionEvent e) {
                    //                    close();
                }
            });
            if (OS.isMac()) add(closeItem);
            add(saveItem = new JMenuItem(new CustomAction("Save", KeyEvent.VK_S, ks(KeyEvent.VK_S, mod)) {
                @Override
                void action(ActionEvent e) {
                    if (hudEditor.canvas.getElements().isEmpty()) return;
                    hudEditor.info(hudEditor.canvas.getElements().getLast().save());
                }
            }));
            saveItem.setEnabled(false);
            add(saveAsItem = new JMenuItem(new CustomAction("Save As...",
                    KeyEvent.VK_A,
                    ks(KeyEvent.VK_S, mod | ActionEvent.SHIFT_MASK)) {
                @Override
                void action(ActionEvent e) {
                }
            }));
            saveAsItem.setEnabled(false);
            add(reloadItem = new JMenuItem(new CustomAction("Revert", KeyEvent.VK_R, ks(KeyEvent.VK_F5)) {
                @Override
                void action(ActionEvent e) {
                    hudEditor.loadAsync(hudEditor.lastLoaded);
                }
            }));
            reloadItem.setEnabled(false);
            exitItem = new JMenuItem(new CustomAction("Exit", KeyEvent.VK_X, ks(KeyEvent.VK_Q, mod)) {
                @Override
                void action(ActionEvent e) {
                    hudEditor.quit();
                }
            });
            if (!OS.isMac()) {
                addSeparator();
                add(closeItem);
                add(exitItem);
            }
        }});
        add(new JMenu("Edit") {{
            setMnemonic(KeyEvent.VK_E);
            add(undoItem = new JMenuItem(new CustomAction("Undo", KeyEvent.VK_U, ks(KeyEvent.VK_Z, mod)) {
                @Override
                void action(ActionEvent e) {
                }
            }));
            undoItem.setEnabled(false);
            // TODO: ctrl + shift + z
            add(redoItem = new JMenuItem(new CustomAction("Redo", KeyEvent.VK_R, ks(KeyEvent.VK_Y, mod)) {
                @Override
                void action(ActionEvent e) {
                }
            }));
            redoItem.setEnabled(false);
            addSeparator();
            add(cutItem = new JMenuItem(new CustomAction("Cut", KeyEvent.VK_T, ks(KeyEvent.VK_X, mod)) {
                @Override
                void action(ActionEvent e) {
                }
            }));
            cutItem.setEnabled(false);
            add(copyItem = new JMenuItem(new CustomAction("Copy", KeyEvent.VK_C, ks(KeyEvent.VK_C, mod)) {
                @Override
                void action(ActionEvent e) {
                }
            }));
            copyItem.setEnabled(false);
            add(pasteItem = new JMenuItem(new CustomAction("Paste", KeyEvent.VK_P, ks(KeyEvent.VK_V, mod)) {
                @Override
                void action(ActionEvent e) {
                }
            }));
            pasteItem.setEnabled(false);
            add(deleteItem = new JMenuItem(new CustomAction("Delete", KeyEvent.VK_D, ks(KeyEvent.VK_DELETE)) {
                @Override
                void action(ActionEvent e) {
                    hudEditor.canvas.removeElements(hudEditor.canvas.getSelected());
                }
            }));
            addSeparator();
            add(selectAllItem = new JMenuItem(new CustomAction("Select All", KeyEvent.VK_A, ks(KeyEvent.VK_A, mod)) {
                @Override
                void action(ActionEvent e) {
                    for (Element elem : hudEditor.canvas.getElements()) {
                        hudEditor.canvas.select(elem);
                    }
                }
            }));
            addSeparator();
            if (!OS.isMac()) {
                add(preferencesItem = new JMenuItem(new CustomAction("Preferences", KeyEvent.VK_E) {
                    @Override
                    void action(ActionEvent e) {
                        hudEditor.preferences();
                    }
                }));
            }
        }});
        add(new JMenu("View") {{
            setMnemonic(KeyEvent.VK_V);
            add(resolutionItem = new JMenuItem(new CustomAction("Change Resolution",
                    KeyEvent.VK_R,
                    ks(KeyEvent.VK_R, mod)) {
                @Override
                void action(ActionEvent e) {
                    hudEditor.changeResolution();
                }
            }));
            resolutionItem.setEnabled(false);
            add(previewItem = new JMenuItem(new CustomAction("Full Screen Preview",
                    KeyEvent.VK_F,
                    ks(KeyEvent.VK_F11)) {
                boolean fullscreen;

                @Override
                void action(ActionEvent e) {
                    hudEditor.dispose();
                    hudEditor.setUndecorated(!fullscreen);
                    hudEditor.setExtendedState(fullscreen ? Frame.NORMAL : Frame.MAXIMIZED_BOTH);
                    hudEditor.setVisible(true);
                    hudEditor.setJMenuBar(hudEditor.editorMenuBar);
                    hudEditor.pack();
                    hudEditor.toFront();
                    fullscreen = !fullscreen;
                }
            }));
            addSeparator();
            JMenuItem viewItem1 = new JMenuItem("Main Menu");
            viewItem1.setEnabled(false);
            add(viewItem1);
            JMenuItem viewItem2 = new JMenuItem("In-game (Health and ammo)");
            viewItem2.setEnabled(false);
            add(viewItem2);
            JMenuItem viewItem3 = new JMenuItem("Scoreboard");
            viewItem3.setEnabled(false);
            add(viewItem3);
            JMenuItem viewItem4 = new JMenuItem("CTF HUD");
            viewItem4.setEnabled(false);
            add(viewItem4);
        }});
        if (!OS.isMac()) {
            add(new JMenu("Help") {{
                setMnemonic(KeyEvent.VK_H);
                add(aboutItem = new JMenuItem(new CustomAction("About", KeyEvent.VK_A) {
                    @Override
                    void action(ActionEvent e) {
                        hudEditor.about();
                    }
                }));
            }});
        }
    }

    KeyStroke ks(int keyCode, int modifiers) {
        return KeyStroke.getKeyStroke(keyCode, modifiers);
    }

    KeyStroke ks(int keyCode) {
        return ks(keyCode, 0);
    }
}

File: src/main/kotlin/com/timepath/hl2/hudeditor/EditorMenuBar.kt
package com.timepath.hl2.hudeditor

import com.timepath.plaf.OS
import java.awt.Frame
import java.awt.Toolkit
import java.awt.event.ActionEvent
import java.awt.event.KeyEvent
import javax.swing.JMenu
import javax.swing.JMenuBar
import javax.swing.JMenuItem
import javax.swing.KeyStroke

class EditorMenuBar(var hudEditor: HUDEditor) : JMenuBar() {
    var reloadItem = JMenuItem(object : CustomAction("Revert", KeyEvent.VK_R, ks(KeyEvent.VK_F5)) {
        override fun action(e: ActionEvent) {
            hudEditor.lastLoaded?.let {
                hudEditor.loadAsync(it)
            }
        }
    })
    val mod = Toolkit.getDefaultToolkit().getMenuShortcutKeyMask()

    init {
        add(object : JMenu(Main.getString("File")) {
            init {
                setMnemonic(KeyEvent.VK_F)
                var newItem = JMenuItem(object : CustomAction(Main.getString("New"), KeyEvent.VK_N, ks(KeyEvent.VK_N, mod)) {
                    override fun action(e: ActionEvent) {
                        throw UnsupportedOperationException("Not supported yet.")
                    }
                })
                add(newItem)
                newItem.setEnabled(false)
                var openItem = JMenuItem(object : CustomAction("Open", KeyEvent.VK_O, ks(KeyEvent.VK_O, mod)) {
                    override fun action(e: ActionEvent) {
                        hudEditor.locateHudDirectory()
                    }
                })
                add(openItem)
                addSeparator()
                var closeItem = JMenuItem(object : CustomAction("Close", KeyEvent.VK_C, ks(KeyEvent.VK_W, mod)) {
                    override fun action(e: ActionEvent) {
                        //                    close();
                    }
                })
                if (OS.isMac()) add(closeItem)
                var saveItem = JMenuItem(object : CustomAction("Save", KeyEvent.VK_S, ks(KeyEvent.VK_S, mod)) {
                    override fun action(e: ActionEvent) {
                        if (hudEditor.canvas!!.elements.isEmpty()) return
                        hudEditor.info(hudEditor.canvas!!.elements.getLast().save())
                    }
                })
                add(saveItem)
                saveItem.setEnabled(false)
                var saveAsItem = JMenuItem(object : CustomAction("Save As...", KeyEvent.VK_A, ks(KeyEvent.VK_S, mod or ActionEvent.SHIFT_MASK)) {
                    override fun action(e: ActionEvent) {
                    }
                })
                add(saveAsItem)
                saveAsItem.setEnabled(false)

                add(reloadItem)
                reloadItem.setEnabled(false)
                var exitItem = JMenuItem(object : CustomAction("Exit", KeyEvent.VK_X, ks(KeyEvent.VK_Q, mod)) {
                    override fun action(e: ActionEvent) {
                        hudEditor.quit()
                    }
                })
                if (!OS.isMac()) {
                    addSeparator()
                    add(closeItem)
                    add(exitItem)
                }
            }
        })
        add(object : JMenu("Edit") {
            init {
                setMnemonic(KeyEvent.VK_E)
                var undoItem = JMenuItem(object : CustomAction("Undo", KeyEvent.VK_U, ks(KeyEvent.VK_Z, mod)) {
                    override fun action(e: ActionEvent) {
                    }
                })
                add(undoItem)
                undoItem.setEnabled(false)
                // TODO: ctrl + shift + z
                var redoItem = JMenuItem(object : CustomAction("Redo", KeyEvent.VK_R, ks(KeyEvent.VK_Y, mod)) {
                    override fun action(e: ActionEvent) {
                    }
                })
                add(redoItem)
                redoItem.setEnabled(false)
                addSeparator()
                var cutItem = JMenuItem(object : CustomAction("Cut", KeyEvent.VK_T, ks(KeyEvent.VK_X, mod)) {
                    override fun action(e: ActionEvent) {
                    }
                })
                add(cutItem)
                cutItem.setEnabled(false)
                var copyItem = JMenuItem(object : CustomAction("Copy", KeyEvent.VK_C, ks(KeyEvent.VK_C, mod)) {
                    override fun action(e: ActionEvent) {
                    }
                })
                add(copyItem)
                copyItem.setEnabled(false)
                var pasteItem = JMenuItem(object : CustomAction("Paste", KeyEvent.VK_P, ks(KeyEvent.VK_V, mod)) {
                    override fun action(e: ActionEvent) {
                    }
                })
                add(pasteItem)
                pasteItem.setEnabled(false)
                var deleteItem = JMenuItem(object : CustomAction("Delete", KeyEvent.VK_D, ks(KeyEvent.VK_DELETE)) {
                    override fun action(e: ActionEvent) {
                        hudEditor.canvas!!.r.removeElements(hudEditor.canvas!!.r.selected)
                    }
                })
                add(deleteItem)
                addSeparator()
                var selectAllItem = JMenuItem(object : CustomAction("Select All", KeyEvent.VK_A, ks(KeyEvent.VK_A, mod)) {
                    override fun action(e: ActionEvent) {
                        for (elem in hudEditor.canvas!!.elements) {
                            hudEditor.canvas!!.r.select(elem)
                        }
                    }
                })
                add(selectAllItem)
                addSeparator()
                if (!OS.isMac()) {
                    var preferencesItem = JMenuItem(object : CustomAction("Preferences", KeyEvent.VK_E) {
                        override fun action(e: ActionEvent) {
                            hudEditor.preferences()
                        }
                    })
                    add(preferencesItem)
                }
            }
        })
        add(object : JMenu("View") {
            init {
                setMnemonic(KeyEvent.VK_V)
                var resolutionItem = JMenuItem(object : CustomAction("Change Resolution", KeyEvent.VK_R, ks(KeyEvent.VK_R, mod)) {
                    override fun action(e: ActionEvent) {
                        hudEditor.changeResolution()
                    }
                })
                add(resolutionItem)
                resolutionItem.setEnabled(false)
                var previewItem = JMenuItem(object : CustomAction("Full Screen Preview", KeyEvent.VK_F, ks(KeyEvent.VK_F11)) {
                    var fullscreen: Boolean = false

                    override fun action(e: ActionEvent) {
                        hudEditor.dispose()
                        hudEditor.setUndecorated(!fullscreen)
                        hudEditor.setExtendedState(if (fullscreen) Frame.NORMAL else Frame.MAXIMIZED_BOTH)
                        hudEditor.setVisible(true)
                        hudEditor.setJMenuBar(hudEditor.editorMenuBar)
                        hudEditor.pack()
                        hudEditor.toFront()
                        fullscreen = !fullscreen
                    }
                })
                add(previewItem)
                addSeparator()
                val viewItem1 = JMenuItem("Main Menu")
                viewItem1.setEnabled(false)
                add(viewItem1)
                val viewItem2 = JMenuItem("In-game (Health and ammo)")
                viewItem2.setEnabled(false)
                add(viewItem2)
                val viewItem3 = JMenuItem("Scoreboard")
                viewItem3.setEnabled(false)
                add(viewItem3)
                val viewItem4 = JMenuItem("CTF HUD")
                viewItem4.setEnabled(false)
                add(viewItem4)
            }
        })
        if (!OS.isMac()) {
            add(object : JMenu("Help") {
                init {
                    setMnemonic(KeyEvent.VK_H)
                    var aboutItem = JMenuItem(object : CustomAction("About", KeyEvent.VK_A) {
                        override fun action(e: ActionEvent) {
                            hudEditor.about()
                        }
                    })
                    add(aboutItem)
                }
            })
        }
    }

    fun ks(keyCode: Int, modifiers: Int): KeyStroke {
        return KeyStroke.getKeyStroke(keyCode, modifiers)
    }

    fun ks(keyCode: Int): KeyStroke {
        return ks(keyCode, 0)
    }
}
